### DISCUSSION TITLE
Update the upstream proxy without SSL interception

### DISCUSSION THREAD

**Discussion by zioalex** [mitmproxy/mitmproxy#5447] [Q&A] [answered]
Hi dear community,
I am trying to set the upstream proxy dynamically based on some rules.
I modified the the example available here https://github.com/mitmproxy/mitmproxy/blob/main/examples/contrib/change_upstream_proxy.py to do it. And it worked.

Later on I noticed that on some downloads, specifically in git fetch, mitmproxy sucks all the resources and it is very slow.

I then thought to exclude the SSL interception with this code
```
class IgnoreTLS:
    def __init__(self) -> None:
        pass

    def tls_clienthello(self, data: mitmproxy.proxy.layers.tls.ClientHelloData):
        '''
        ignore some tls event
        '''
            
        server = re.findall(r'([\w\-_]+(?:(?:\.[\w\-_]+)+))', str(data.context.server))        
        
        ctx.log(f"Server found: {server[0]}")


        if re.match(exclude_k8s_mngt, server[0]):
            ctx.log(f"TLS passthrough: {server[0]}")
            data.ignore_connection = True

addons = [
    IgnoreTLS()
]
```

Now the 2 operations, ignore the TLS and change the upstream proxy dynamically, work only one at the time.

Can you help me to get the upstream proxy set dynamically and exclude the TLS interception for some or all hosts?

Thanks,
Alessandro

**Comment by alber7rp** [comment, marked as answer]
Hello @zioalex ,
I had the same problem as you. I was able to set dinamically the upstream server handling the "CONNECT" event from the client. You should to handle it and the set the chained proxy address:

``` python
def server_connect(data):
    print("CONNECT command")
    data.server.address = ('1.2.3.4', 8080) #Address of the upstream proxy
```

**Reply by zioalex** [reply]
Thank you @alber7rp,
will try if.
